plenum/server/consensus/checkpoint_service.py

```diff
@@ -126,215 +99,108 @@ def process_ordered(self, ordered: Ordered):
         for batch_id in reversed(self._data.preprepared):
             if batch_id.pp_seq_no == ordered.ppSeqNo:
                 self._add_to_checkpoint(batch_id.pp_seq_no,
-                                        batch_id.pp_digest,
-                                        ordered.ledgerId,
                                         batch_id.view_no,
                                         ordered.auditTxnRootHash)
                 return
         raise LogicError("CheckpointService | Can't process Ordered msg because "
                          "ppSeqNo {} not in preprepared".format(ordered.ppSeqNo))
 
-    def _start_catchup_if_needed(self):
-        stashed_checkpoint_ends = self._stashed_checkpoints_with_quorum()
-        lag_in_checkpoints = len(stashed_checkpoint_ends)
-        if self._checkpoint_state:
-            (s, e) = firstKey(self._checkpoint_state)
-            # If the first stored own checkpoint has a not aligned lower bound
-            # (this means that it was started after a catch-up), is complete
-            # and there is a quorumed stashed checkpoint from other replicas
-            # with the same end then don't include this stashed checkpoint
-            # into the lag
-            if s % self._config.CHK_FREQ != 0 \
-                    and self._checkpoint_state[(s, e)].seqNo == e \
-                    and e in stashed_checkpoint_ends:
-                lag_in_checkpoints -= 1
-        is_stashed_enough = \
-            lag_in_checkpoints > self.STASHED_CHECKPOINTS_BEFORE_CATCHUP
-        if not is_stashed_enough:
+    def _start_catchup_if_needed(self, key: CheckpointKey):
+        if self._have_own_checkpoint(key):
+            return
+
+        unknown_stabilized = self._unknown_stabilized_checkpoints()
+        lag_in_checkpoints = len(unknown_stabilized)
+        if lag_in_checkpoints <= self.STASHED_CHECKPOINTS_BEFORE_CATCHUP:
             return
 
+        last_key = sorted(unknown_stabilized, key=lambda v: (v.view_no, v.pp_seq_no))[-1]
+
         if self.is_master:
-            self._logger.display(
-                '{} has lagged for {} checkpoints so updating watermarks to {}'.format(
-                    self, lag_in_checkpoints, stashed_checkpoint_ends[-1]))
-            self.set_watermarks(low_watermark=stashed_checkpoint_ends[-1])
+            # TODO: This code doesn't seem to be needed, but it was there. Leaving just in case
+            #  tests explain why it was really needed.
+            # self._logger.display(
+            #     '{} has lagged for {} checkpoints so updating watermarks to {}'.format(
+            #         self, lag_in_checkpoints, last_key.pp_seq_no))
+            # self.set_watermarks(low_watermark=last_key.pp_seq_no)
```

Comment:
I looked at the history, and I believe this is not needed, so it should be fine to remove it.
- initially this was added to correctly move watermarks on backup instances (it was not under `if self.is_master`)
- later on the logic of 'fake catch-up' (more steps than just moving watermarks) has been implemented, and these lines were moved under `if self.is_master` without an actual need for this. 
- since we do not order anything during catchup and move watermarks after catch-up is finished, this logic can be removed.

Reply:
Okay, I'll remove it in next PR then after final cleanup